mypy/fastparse.py

```diff
@@ -271,8 +276,23 @@ def visit(self, node: Optional[AST]) -> Any:
             method = 'visit_' + node.__class__.__name__
             visitor = getattr(self, method)
             self.visitor_cache[typeobj] = visitor
+        # In Python 3.8, we can expand the scope of ignores to a whole expression:
+        if sys.version_info >= (3, 8) and isinstance(node, ast3.expr):
+            self.scope_ignores(node)
         return visitor(node)
 
+    def scope_ignores(self, node: ast3.expr) -> None:
+        end_lineno = getattr(node, "end_lineno", None)
+        if end_lineno is None:
+            return
+        node_lines = range(node.lineno, end_lineno + 1)
+        # Check to see if this expression overlaps with any "type: ignore" comments.
+        # If so, take the first one and grow its scope to cover the whole node:
+        for line in node_lines:
+            if line in self.type_ignore_comments:
+                self.type_ignores.update(dict.fromkeys(node_lines, line))
```

Comment:
It seems to me now there is a flaw in this logic. Imagine this test case:
```python
def func(x: int) -> None: ...
lst = [func(1),
       func('no'),  # type: ignore
       func('yes'),
       func(2)]
```
I think only the error for `func('no')` should be ignored, while we still should show the error for `func('yes')`.
It looks like the actual solution should be a bit more complex.

In some sense the logic should be reversed, for every error you should check whether there is a `# type: ignore` comment in the span of error context expression for that particular error.

I am not sure what is the best way, but maybe we would need to copy the end position for every expression in the generated mypy AST.

Reply:
I agree now, when you put it like that. Another clear example would be:

```py
def f(a: int): ...
[f("error"), f(
    "ignored"  # type: ignore
    ), f("error")]
```

...which isn’t really possible by predefining the line scope of each ignore. Really, it’s expressions that have scopes, not the ignores. This fix is exactly backwards.

I think that having the expression context grow an `end_lineno` slot and checking for ignores when reporting is both correct and more efficient. I’ll back out the current implementation and shift to that approach, if that seems right to you @ilevkivskyi!

Reply:
OK, go ahead, and let me know when the PR is ready for another review.